### PR TITLE
Added Jenkins Pipeline 

### PR DESCRIPTION
**Pull request by nicolehaugen79**
This PR includes a script for the Jenkins pipeline.  I would appreciate having others try this out to see if it's successful for you.  I also updated the README with basic steps.

The other change I made as part of this is to make the ingress controller configurable so that it is only hooked up for the production version of the app and NOT the test version.  

Last but not least, I updated the checkout service docker file with feedback we received after meeting with msimons (and others) - for reference, here's the explanation behind these changes:

[msimons] In the checkoutServer Dockerfile...the benefit of this pattern is that if you are doing dev work, you won’t have to incur the cost of restoring each time you rebuild the image.  Additionally I would recommend against using `dotnet run` as your ENTRYPOINT or CMD.  `run` has to do some additional checks to ensure the source doesn’t need to be rebuilt.  This adds unnecessary overhead each time the container is started.

### CODE REVIEW

**Comment by chlowell** [comment]
What syntax trouble are you having in `services.yaml`? I think this would work:
```go
{{- if .Values.useIngress }}
  type: NodePort
{{- end }}
```

**Comment by nicolehaugen79** [comment]
I was getting an error when deploying because the example I originally followed showed having a hyphen at the beginning and end of the if-condition.  It seems to complain about that in some places and not others.  I changed the if-condition to exactly what you have above and now it works.  Thanks!

File: README.md [hunk 1]
```diff
@@ -150,3 +150,104 @@ honest-deer-apigateway   2         2         2            2           10m
...
+
+4. Run the pipeline script.
+There are 2 ways to trigger the pipeline script to run.
+* You can push a change to GitHub.  Note that this requires you to add a webhook to Jenkins within GitHub's settings.
+  In GitHub, under the repo's settings, select Integrations & services.  Choose to Add a service and select Jenkins (GitHub plugin).  In the Jenkins hook url field,
+  specify the url to your Jenkins VM followed by "/github-webhook".  For example: http://<IP>:8080/github-webhook/.  Ensure that the Active checkbox is checked.
+* Or, within Jenkins, you can click the Build Now link.
+
+## AAD Setup
+The app's authentication service is implemented as part of the API Gateway and supports both basic email and facebook authentication by using Azure AAD B2C.  The API Gateway acts as the primary entry point into th eserver by providing a wrapper over all calls to the microservices' endpoints.  The advantage of this approach is:
```
Comment (chlowell) [inline README.md]: `th eserver`

File: jenkins-scripts/ci-pipeline.groovy [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+   // Name of the AAD tenant used for the app's B2C authentication, e.g. <tenantname>.onmicrosoft.com
+   env.AAD_TENANT = ''
+   
+   // URL for the internal API Gateway service that the integration tests run against
+   env.TEST_APP_URL = "http://localhost:8001/api/v1/proxy/namespaces/default/services/$TEST_RELEASE-apigateway"
+   
+   stage('Preparation') { 
+    
+      // Get some code from a GitHub repository
+      git 'https://github.com/nicolehaugen79/StickerApp.git'
```
Comment (chlowell) [inline jenkins-scripts/ci-pipeline.groovy]: Should we use `Glimpse/StickerApp` here?

File: README.md [hunk 3]
```diff
@@ -150,3 +150,104 @@ honest-deer-apigateway   2         2         2            2           10m
...
+2. Add credentials required by the pipeline script.
+Go to Credentials->System, create the following credentials using the IDs outlined below:
+* Kind: Username with password, ID: AAD 
+  - Stores the AAD client and secret needed by the Sticker App to authenticate users of the app.
+* Kind: Username with password, ID: DOCKER
+  - Stores the Azure Container Registry user id and secret needed to push images to the registry.
+* Kind: Secret text, ID: ACR
+  - Stores the Docker secret that is needed by Kunernetes to deploy images from the registry.
+
+3. Create the pipeline and add the script.
```
Comment (xiangyan99) [inline README.md]: Looks like you missed the steps to install helm.

File: README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Sticker E-Commerce Demo App
 
-**Running locally:**
+## Running Locally
 
 ```console
 $ docker-compose -f docker-compose.dev.yml up -d
```
Comment (chlowell) [inline README.md]: We need to add something about the AAD secrets, which in the compose environment come from `apigateway/debug.env`. `apigateway` crashes when these aren't set.

File: jenkins-scripts/ci-pipeline.groovy [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            else
+                helm install stickerapp --name $TEST_RELEASE --wait --set useIngress="false" --set azureActiveDirectory.clientId="$AAD_CLIENTID" --set azureActiveDirectory.clientSecret="$AAD_SECRET" --set azureActiveDirectory.destroySessionUrl="https://login.microsoftonline.com/$AAD_TENANT/oauth2/v2.0/logout?p=B2C_1_SignIn&post_logout_redirect_uri=https://$INGRESS_IP" --set azureActiveDirectory.redirectUrl="https://$INGRESS_IP/users/auth/return" --set azureActiveDirectory.tenant="$AAD_TENANT" --set registry="$ACR_NAME" --set dockercfg="$DOCKER_SECRET" --set imageTag="$BUILD_NUMBER" --set kafkaBroker="$KAFKA_BROKER" --set zookeeperConnect="$ZK_CONNECT"
+            fi'''
+            
+            // Helm's --wait command isn't sufficient to determine if deployment is ready, so check to see if the Session's deployment is available before running tests
+            env.TEST_SESSION_NAME = env.TEST_RELEASE + "-session"
+            sh '''
+            for i in {1 2 3 4 5 6 7 8 9 10 11 12 13 14 15}
```
Comment (nicolehaugen79) [inline jenkins-scripts/ci-pipeline.groovy]: I had tried to simplify this and unfortunately the only way I could get this loop to work through Jenkins is how I have it.

File: README.md (hunk 3 above)
Comment (nicolehaugen79) [inline README.md]: Good question...I will investigate this further to see if I can find a way.  For now though, I would like to go ahead with this check-in and work on that afterward.

File: README.md [hunk 6]
```diff
@@ -150,3 +150,104 @@ honest-deer-apigateway   2         2         2            2           10m
...
+2. [Register the app](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-app-registration)
+* Note that the Reply URL will need to include the external IP address of the cluster's ingress controller, e.g. "https://<ingress controller IP>/users/auth/return"
+
+3. [Create Sign Up\Sign In policies](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-reference-policies#create-a-sign-up-policy)
+
+In addition, refer to the 'Create an application' and 'Create your policies' sections that are included [here](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-devquickstarts-web-node)
+
+4. [Configure Facebook with your AAD tenant](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-setup-fb-app)
+
+Note that there are steps missing from this - the signup/login policies need to be added the new Facebook provider too.
```
Comment (nicolehaugen79) [inline README.md]: I removed the line that says "Note that there are steps missing..." - After I wrote this, I submitted a PR to the docs to have these steps added: https://github.com/MicrosoftDocs/azure-docs/pull/2210

File: jenkins-scripts/ci-pipeline.groovy (hunk 2 above)
Comment (nicolehaugen79) [inline jenkins-scripts/ci-pipeline.groovy]: ha...yes...thanks for catching

**Comment by nicolehaugen79** [comment]
Incorporated all review feedback with the exception of the credential scripting for the CI pipeline (I have decided to wait on that for now).

**Pull request merged**